### DISCUSSION TITLE
Add header authentication

### DISCUSSION THREAD

**Discussion by PhilippMundhenk** [mealie-recipes/mealie#2607] [Feature Request]
To avoid having to tackle the more complex OpenID Connect stuff, reverse proxy/header authentication could be used instead / as a stop gap. Basically you run the application behind a reverse proxy, and the proxy authenticates the user and adds `Remote-User` and `Remote-Groups` (if applicable) headers to the request to tell the app which user is logged in, the app often has a "trusted proxies" configuration which prevents other sources from using this to bypass auth.

_Originally posted by @m00nwtchr in https://github.com/mealie-recipes/mealie/issues/1197#issuecomment-1280044865_

**Comment by gesquive** [comment]
This would be amazing. If you are looking for examples of how to implement this, [Grafana](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/auth-proxy/) & [GitTea](https://docs.gitea.com/next/usage/authentication#reverse-proxy) both have solid implementations.

**Comment by GieltjE** [comment]
This would indeed be helpfull, all my services are being authenticated by apache, which allows easy usage from my local network and secure remote access.

**Comment by max-tet** [comment]
I also would find this feature extremely helpful! Just to give a little more detail, the perfect proxy auth feature for me would work like this:

1. Start the app with additional environment variables:

* containing the name of the initial admin user (e.g. admin=admin_user)
* enabling proxy auth (e.g. proxy_auth=true)
* setting the key of the http header that contains the username (e.g. auth_header=X-Authenticated-User)

2. Configure the reverse proxy to authenticate incoming requests in any way you like.
3. Let the reverse proxy set X-Authenticated-User to the authenticated username on every request.
4. The app treats the requests as if they belong to the appropriate user session.
5. Bonus: if the app does not know the username, it creates a new user with that name.

**Comment by aroberts** [comment]
@cmintey do you know if your work landing OIDC auth makes this project any easier? I explored this a bit but nuxtjs is pretty opaque to me, unfortunately, and I inferred from trawling the docs and a comment you made somewhere that there's nothing built into nuxt's auth package to really facilitate here. Is there machinery in place now (or perhaps also leveraging parts of #2040) that might make this project within reach?

**Comment by vexofp** [comment]
Here is a branch in which I have made an initial attempt at implementing this: https://github.com/mealie-recipes/mealie/compare/mealie-next...vexofp:mealie:header-auth . It is still missing docs and tests and a few other things before being ready for a PR, but I would appreciate any initial feedback as to the high-level approach!

It adds settings by which you can specify trusted headers that will contain the username/email/full name/role, then automatically creates or updates and logs in the indicated user. I am using it successfully behind NGINX, with the header values populated from the client's mTLS certificate.

**Reply by GieltjE** [reply]
Is there a quick 'n dirty guide on what variables/configs to set?

**Reply by vexofp** [reply]
At minimum you need to set `HEADER_AUTH_ENABLED` to true and `HEADER_AUTH_USERNAME` to the name of the trusted HTTP header that will contain the authenticated user's username. New users will be created automatically as needed (similar to LDAP auth). You can optionally also set `
HEADER_AUTH_FULL_NAME`, `HEADER_AUTH_EMAIL`, and `HEADER_AUTH_ROLE` to other header names that will populate/update the corresponding fields. (With "admin" role granting `is_admin` and anything else a normal user.)
